### ISSUE TITLE
Migrate http based sinks to use `HttpSink`

### ISSUE THREAD

**Issue by LucioFranco** [timberio/vector#1659]
As of #1641 we now have a new `HttpSink` trait that unifies how we build our http sinks. We need to follow up this work by migrating the rest of our http based sinks to be implemented similarly to the current `http` sink.

- [x] `clickhouse`
- [x] `datadog_metrics`
- [x] `elasticsearch`
- [ ] `gcp_cloud_storage`
- [x] `gcp_pubsub`
- [x] `gcp_stackdriver_logging`
- [x] `new_relic_logs`
- [x] `splunk_hec`

We will need to also support adding partitioned sinks, this should be possible but to keep the type system happy we will need to implement a form of `Either` that can wrap our `BatchedSink` and `PartitonBatchedsink`.

**Comment by bruceg** [comment]
Both the `gcp_pubsub` and `gcp_stackdriver_logging` sinks use JSON over HTTP, so this would likely be applicable to them as well.

**Comment by LucioFranco** [comment]
Updated the description :+1:

**Comment by LucioFranco** [comment]
Done with #2075

**Comment by LucioFranco** [comment]
`gcp_cloud_storage` was not worth it to migrate but the rest are done!

**Comment by binarylogic** [comment]
Thanks. Mind sharing details why?

**Comment by LucioFranco** [comment]
@binarylogic reviewing a lot of the partition based sinks the complexity that they have is worth it for what they provide. So basically there is no benefit to use httpsink trait since you still need the extra logic surrounding it. We also use this type of a code a lot less so in the end it makes sense to leave it how it is but swap out the actual sink types. 

TL;DR: in my experience Service is the right abstraction here rather than HttpSink.
